### PR TITLE
Wire contexts to Autoscaling controllers

### PR DESCRIPTION
**Pull request by damemi**
#### What type of PR is this?
/kind design

#### What this PR does / why we need it:
This wires a context to Autoscaling controllers in kube-controller-manager, so that KCM can eventually release on cancel.

This originally came from PR https://github.com/kubernetes/kubernetes/pull/101379, which builds on top of https://github.com/kubernetes/kubernetes/pull/101125

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/podautoscaler/horizontal_test.go [hunk 1]
```diff
@@ -747,7 +748,7 @@ func (tc *testCase) runTestWithController(t *testing.T, hpaController *Horizonta
 	stop := make(chan struct{})
 	defer close(stop)
 	informerFactory.Start(stop)
-	go hpaController.Run(stop)
+	go hpaController.Run(context.TODO())
```
Comment (damemi) [inline pkg/controller/podautoscaler/horizontal_test.go:751]: I think the failure is related to this, it seems to be panicking on the calls related to here:

```
E1004 19:10:12.992779   52130 runtime.go:76] Observed a panic: Fail in goroutine after TestUpscaleCap has completed
...
created by k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).Run
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:177 +0x4bc
```

which is a call to `wait.UntilWithContext`, so maybe `context.TODO()` doesn't work for that? Or I just be using the same context for this and the stop channel

**Review by soltysh — COMMENTED** [review]
This mostly looks good, although you have one leftover in https://github.com/kubernetes/kubernetes/blob/79ee735badfe457136ee0da51f3783628898a137/pkg/controller/podautoscaler/metrics/client.go#L68
and the unit failure looks legit too

**Review by soltysh — CHANGES_REQUESTED** [review]

**Review by soltysh — APPROVED** [review]
/triage accepted
/priority backlog
/lgtm
/approve

**Pull request merged**